src/tools/thmgen/thmgen.cs

```diff
@@ -0,0 +1,165 @@
+
+            // load the theme file and find all the controls and pages with a Name attribute
+            var doc = XDocument.Load(this.commandLine.InputFile);
+            var named = doc.Descendants().Elements().Where(xelem => xelem.Attribute("Name") != null);
+            var controls = named.Where(xelem => xelem.Name != this.NameSpace + "Page");
+            var pages = named.Where(xelem => xelem.Name == this.NameSpace + "Page");
+
+            // theme.h
+            var templateControlsDeclaration = GetTemplateFromResource("WixToolset.Tools.Templates.Controls.h");
+            var templatePagesDeclaration = GetTemplateFromResource("WixToolset.Tools.Templates.Pages.h");
+
+            var controlsDeclaration = String.Join(",\n", GetDeclaration(controls));
+            var pagesDeclaration = String.Join(",\n", GetDeclaration(pages));
+
+            File.WriteAllText(outputHeaderFile, String.Format(templateControlsDeclaration, prefix, controlsDeclaration));
+            File.AppendAllText(outputHeaderFile, String.Format(templatePagesDeclaration, prefix, pagesDeclaration));
+
+            // theme.cpp
+            var templateControlsInitialization = GetTemplateFromResource("WixToolset.Tools.Templates.Controls.cpp");
+            var templatePagesInitializations = GetTemplateFromResource("WixToolset.Tools.Templates.Pages.cpp");
+
+            var controlsInitialization = String.Join(",\n", GetControlsInitialization(controls, prefix));
+            var pagesInitialization = String.Join(",\n", GetPagesInitialization(pages));
+
+            File.WriteAllText(outputModuleFile, String.Format(templateControlsInitialization, prefix, controlsInitialization, pch));
+            File.AppendAllText(outputModuleFile, String.Format(templatePagesInitializations, prefix, pagesInitialization));
+        }
+
+        private static string GetTemplateFromResource(string resourceName)
+        {
+            using (var stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(resourceName))
+            using (var reader = new StreamReader(stream))
+            {
+                return reader.ReadToEnd();
+            }
+        }
+
+        private static IEnumerable<string> GetDeclaration(IEnumerable<XElement> elems)
+        {
+            return elems.Select(elem => elem.Attribute("Name").Value);
+        }
+
+        private static IEnumerable<string> GetPagesInitialization(IEnumerable<XElement> pages)
+        {
+            return pages.Select(page =>
+            {
+                var name = page.Attribute("Name").Value;
+                return String.Format("L\"{0}\"", name);
+            });
+        }
+
+        private static IEnumerable<string> GetControlsInitialization(IEnumerable<XElement> controls, string prefix)
+        {
+            return controls.Select(control =>
+            {
+                var name = control.Attribute("Name").Value;
+                return String.Format("{{ {0}Controls::{1}, L\"{1}\" }}", prefix, name);
```

Comment:
Investigate whether we need to block out ids for things like IDCANCEL that Windows will send.